File: POO/Constructores_Clases/Abstract class/JAVA/Dispositivo.java
public abstract class Dispositivo {
    protected String marca;
    protected String modelo;

    public abstract void encender();
    public abstract void apagar();
    public abstract void mostrar();
}

class Tablet extends Dispositivo {
    private double pulgadas;
    private double precio;
    private boolean encendido;

    // Constructor 1: Solo marca y modelo
    public Tablet(String marca, String modelo) {
        this.marca = marca;
        this.modelo = modelo;
        this.pulgadas = 10.0;
        this.precio = 0.0;
        this.encendido = false;
    }

    // Constructor 2: Todos los parámetros
    public Tablet(String marca, String modelo, double pulgadas, double precio, boolean encendido) {
        this.marca = marca;
        this.modelo = modelo;
        this.pulgadas = pulgadas;
        this.precio = precio;
        this.encendido = encendido;
    }

    @Override
    public void encender() {
        encendido = true;
        System.out.println("Tablet encendida");
    }

    @Override
    public void apagar() {
        encendido = false;
        System.out.println("Tablet apagada");
    }

    @Override
    public void mostrar() {
        String estado = encendido ? "Encendida" : "Apagada";
        System.out.println(marca + " " + modelo + " - " + pulgadas + " pulgadas");
        System.out.printf("Precio: $%.2f%n", precio);
        System.out.println("Estado: " + estado);
    }

    public static void main(String[] args) {
        Dispositivo tablet1 = new Tablet("Samsung", "Tab S7");
        Dispositivo tablet2 = new Tablet("Apple", "iPad Pro", 12.9, 1200.00, true);

        tablet1.mostrar();
        tablet2.mostrar();

        tablet1.encender();
        tablet1.mostrar();
    }
}

File: POO/Constructores_Clases/Abstract class/GO/Dispositivo.go
package main

import "fmt"

type Dispositivo interface {
    Encender() string
    Apagar() string
    Estado() string
    Mostrar()
}

type Computadora struct {
    marca   string
    modelo  string
    encendida bool
    precio  float64
}

// Constructor 1: Solo marca y modelo
func NewComputadora(marca, modelo string) *Computadora {
    return &Computadora{
        marca:    marca,
        modelo:   modelo,
        encendida: false,
        precio:   0.0,
    }
}

// Constructor 2: Todos los parámetros
func NewComputadoraCompleta(marca, modelo string, precio float64, encendida bool) *Computadora {
    return &Computadora{
        marca:    marca,
        modelo:   modelo,
        precio:   precio,
        encendida: encendida,
    }
}

func (c *Computadora) Encender() string {
    c.encendida = true
    return "Computadora encendida"
}

func (c *Computadora) Apagar() string {
    c.encendida = false
    return "Computadora apagada"
}

func (c *Computadora) Estado() string {
    if c.encendida {
        return "Encendida"
    }
    return "Apagada"
}

func (c *Computadora) Mostrar() {
    fmt.Printf("%s %s - $%.2f\n", c.marca, c.modelo, c.precio)
    fmt.Printf("Estado: %s\n", c.Estado())
}

func main() {
    var disp1 Dispositivo = NewComputadora("Dell", "XPS")
    var disp2 Dispositivo = NewComputadoraCompleta("Apple", "MacBook", 2000.00, true)
    
    disp1.Mostrar()
    disp2.Mostrar()
}